src/hotspot/share/services/heapDumper.cpp

```diff
@@ -787,6 +789,111 @@ class DumperSupport : AllStatic {
   }
 };
 
+// Hash table of klasses to the klass metadata. This should greatly improve the
+// hash dumping performance. This hash table is supposed to be used by a single
+// thread only.
+//
+class DumperClassCacheTableEntry : public CHeapObj<mtServiceability> {
+  friend class DumperClassCacheTable;
+private:
+  GrowableArray<char> _sigs_start;
+  GrowableArray<int> _offsets;
+  u4 _instance_size;
+  int _entries;
+
+public:
+  DumperClassCacheTableEntry() : _instance_size(0), _entries(0) {};
+
+  int field_count()             { return _entries; }
+  char sig_start(int field_idx) { return _sigs_start.at(field_idx); }
+  int offset(int field_idx)     { return _offsets.at(field_idx); }
+  u4 instance_size()            { return _instance_size; }
+};
+
+class DumperClassCacheTable : public StackObj {
```

Comment:
Technically it's a ValueObj inside a StackObj, so don't really need to have StackObj here. To me, that it's means that you're going to use it as a local variable in a method, when it's really embedded in StackObj.  One can argue all ways on this though, it's just my opinion.

Reply:
Yeah. What is the proper idiom for "I do not want this to be allocated (in C heap or anywhere else) without NMT knowing"? `StackObj` is the closest I can get nowadays, I think. `ValueObj` is gone since JDK 11: https://bugs.openjdk.org/browse/JDK-8199282

Reply:
yes, everything without a base class is a ValueObj and there's a link-time error if you try to call 'new' on it. The ValueObj base class was problematic.  Now it's just omitted.

Reply:
Got it. Dropped `StackObj`. Indeed, if I try to do `new ...Table()`, I get a link-time error.